### ISSUE TITLE
Can we declare local variables as `data`?

### ISSUE THREAD

**Issue by bbbales2** [stan-dev/stanc3#756]
Being able to declare in functions:

```
data real x_r[0];
data int x_i[0];
```

would be handy for the higher order functions as they are now.

Managing what is and is not data makes it difficult to put the higher order functions in user-defined functions: https://discourse.mc-stan.org/t/calling-integrate-1d-reports-parsing-error-parser-expected/19611/9

The usability problems would go away with variadic stuff, but there's still the performance issue (it is very costly in the higher order functions to pass in data things as autodiff types).

**Comment by rok-cesnovar** [comment]
@bbbales2 I think this is already available with stanc3 optimizations: https://github.com/stan-dev/stanc3/issues/544

We really need to split the optimizations up into levels so you can turn on these simple optimizations only.

**Comment by wds15** [comment]
With the optimizations you can make sure that local variables which can be data will become data... but I think @bbbales2 is asking if we can directly declare local variables as data and if there is an issue with that the transpiler should complain about it. Declaring things explicitly as data for local variables does not work at the moment as far as I know.

**Comment by rok-cesnovar** [comment]
Oh, yeah misread that. Thanks.

Yeah, we cant do that right now. I think one of the reasons is it becomes extremely difficult to check. 
Example:

```
functions {
  real foo(real p) {
    data k = p;
    ...
  }
}
```
If this is called in transformed data or GQ this is fine, if used in transformed parameters/model it would fail at the C++ compiler level.

**Comment by bbbales2** [comment]
> With the optimizations you can make sure that local variables which can be data will become data

I think this is better though. The data modifier or whatever has always been a bit weird, and it would be better if the compiler just Did The RIght Thing.

> We really need to split the optimizations up into levels so you can turn on these simple optimizations only.

Could this one be turned on by default?

**Comment by rok-cesnovar** [comment]
Yes, my idea was that this would be O=1 and at some point, when enough testing is done om 01 that becomes default.

**Comment by wds15** [comment]
Is there a way so that I can actually see what magic the optimiser has done?

**Comment by bbbales2** [comment]
> when enough testing is done 01 becomes default.

Is there a way for people to turn these things on now in cmdstanr? I'm a little nervous that if we try to wait for testing I'm just going to forget about this issue until the next person shows up on the forums -- this is a pretty frustrating error to deal with too.

**Comment by rok-cesnovar** [comment]
> Is there a way so that I can actually see what magic the optimiser has done?

The most readable check is reading the debug info of the MIR and optimized MIR with `--debug-optimized-mir-pretty` and `--debug-mir-pretty` to stanc3. You can also observe changes in C++ but that is a bit less friendly to read. 

I have an interface I use in the browser locally, but I pushed it to the web now: https://rok-cesnovar.github.io/stanc3js-demo/optimization.html (sorry for the 1990s style of web pages).

> Is there a way for people to turn these things on now in cmdstanr?

```
library(cmdstanr)
mod <- cmdstan_model("eight_schools.stan", stanc_options = list(O=TRUE))
```
We should split it up for 2.26 if possible so it would be O=1, O=2, ...

**Comment by bbbales2** [comment]
> We should split it up for 2.26 if possible so it would be O=1, O=2, ...

Ooo yes. This makes me think it would be good to start a 2.26 list with dates. With expressions and varmat coming along, it's going to be easy to get lost. We've got like 7 weeks to go?

**Comment by rok-cesnovar** [comment]
> This makes me think it would be good to start a 2.26 list with dates.

+1

> We've got like 7 weeks to go?

Jan 18th RC, Jan25th release or a week earlier.
